### PR TITLE
Bolt Sampler - Neo4j 4.x and cluster features

### PR DESCRIPTION
**Pull request by pecollet**
## Description

Bolt Sampler : 
- added "database" option, required for Neo4j 4.x multi-database support
- added "access mode" option, that allows running against a Neo4j Enterprise causal cluster, with proper query routing using the "neo4://" or "bolt+routing://" URL schemes.
- added "transaction timeout" option, which is a useful parameter when load testing
- fixed error displaying results when "Record Query Results" is enabled, with 4.x Driver (retrieving resultSummary caused early consumption of results).

## Motivation and Context
Enables running JMeter against Neo4j 4.0, 4.1 & 4.2.
Enables running JMeter against Neo4j Enterprise Causal clusters
Enables setting a limit on request execution time.

## How Has This Been Tested?
Tested against : 3.5 standalone Neo4j instance, 3.5 Causal Cluster, 4.2 standalone Neo4j instance, 4.2 Causal Cluster.
Test config : 1 "Bolt Connection Configuration" + 1 "Bolt Request" in a Thread group

## Screenshots (if appropriate):
Jmeter-bolt-sampler

## Types of changes
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java [hunk 1]
```diff
@@ -48,4 +84,27 @@ public boolean isRecordQueryResults() {
     public void setRecordQueryResults(boolean recordQueryResults) {
         this.recordQueryResults = recordQueryResults;
     }
+
+    //returns a SessionConfig object that can be passed to the driver session
+    public SessionConfig getSessionConfig() {
+        SessionConfig.Builder sessionConfigBuilder = SessionConfig.builder()
+                .withDefaultAccessMode(Enum.valueOf(AccessMode.class, getAccessMode()));
+
+        if (database != null && !"".equals(database)) {
```
Comment (FSchumacher) [inline src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:93]: You could probably use `StringUtils#isNotBlank(String)` to make intention more clear.

File: src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltSampler.java [hunk 2]
```diff
@@ -135,12 +146,25 @@ private String request() {
...
     }
 
     private String response(Result result) {
         StringBuilder response = new StringBuilder();
+        List<Record> records;
+        if (isRecordQueryResults()) {
+            //get records already as consume() will exhaust the stream
+            records = result.list();
+        } else {
+            records = null;
```
Comment (FSchumacher) [inline src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltSampler.java:166]: Would an empty List work here, too? (`Collections#emptyList()`) That way, we would not have to deal with a null pointer later.

File: src/protocol/bolt/src/main/resources/org/apache/jmeter/protocol/bolt/sampler/BoltSamplerResources.properties [hunk 3]
```diff
@@ -17,9 +17,16 @@
...
 cypher.displayName=Cypher Statement
 cypher.shortDescription=Cypher Statement
 params.displayName=Params
 params.shortDescription=Params
 recordQueryResults.displayName=Record Query Results
 recordQueryResults.shortDescription=Records the results of queries and displays in listeners such as View Results Tree, this iterates through the entire resultset. Use to debug only.
+accessMode.displayName=Access Mode
+accessMode.shortDescription=Whether it's a READ or WRITE query (affects query routing in clusters)
+database.displayName=Database
+database.shortDescription=Neo4j 4.x : database to query (leave empty for 3.5)
```
Comment (FSchumacher) [inline src/protocol/bolt/src/main/resources/org/apache/jmeter/protocol/bolt/sampler/BoltSamplerResources.properties:30]: There should not be a space in front of the colon in the English translations. You can certainly add it for the French translation :)

File: src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltTestElementBeanInfoSupport.java [hunk 4]
```diff
@@ -45,5 +47,23 @@ protected BoltTestElementBeanInfoSupport(Class<? extends TestBean> beanClass) {
...
+        propertyDescriptor =  property("database", TypeEditor.ComboStringEditor);
+        propertyDescriptor.setValue(DEFAULT, "neo4j");
+
+        propertyDescriptor = property("txTimeout");
+        propertyDescriptor.setValue(NOT_UNDEFINED, Boolean.TRUE);
+        propertyDescriptor.setValue(DEFAULT, "60");
+    }
+
+    private String[] getListAccessModes() {
+        String[] list = {AccessMode.READ.toString(), AccessMode.WRITE.toString()};
```
Comment (FSchumacher) [inline src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/BoltTestElementBeanInfoSupport.java:66]: Could be returned directly.

File: src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java [hunk 5]
```diff
@@ -17,13 +17,49 @@
...
+    public String getAccessMode() {
+        if (accessMode != null) {
+            return accessMode;
+        } else {
+            return "WRITE";
+        }
+    }
+
+    public void setAccessMode(String accessMode) {
+        this.accessMode = accessMode;
```
Comment (FSchumacher) [inline src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:53]: Should we check this against the available access modes from the driver?

**Review by FSchumacher — COMMENTED** [review]
Thanks for the PR. I think it will have to wait until the release 5.4.1, which should happen soon.

File: src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java [hunk 6]
```diff
@@ -17,13 +17,53 @@
...
+    public int getTxTimeout() {
+        return txTimeout;
+    }
+
+    public void setTxTimeout(int txTimeout) {
+        this.txTimeout = txTimeout;
+    }
+
+    public String getAccessMode() {
+        if (accessMode != null) {
```
Comment (ham1) [inline src/protocol/bolt/src/main/java/org/apache/jmeter/protocol/bolt/sampler/AbstractBoltTestElement.java:47]: Consider writing this using a guard clause? i.e.
```
if (accessMode == null) {
    return AccessMode.WRITE.toString();
}
return accessMode;
```

**Comment by FSchumacher** [comment]
Thanks for the PR. Can you test the next nightly or build from trunk?

**Pull request closed without merging**